sql/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java

```diff
@@ -208,4 +263,95 @@ public static Function of(String name, Symbol first, Symbol second, Set<Function
             ImmutableList.of(first, second)
         );
     }
+
+    private static final class IntervalArithmeticScalar extends Scalar<Period, Object> {
+
+        private final FunctionInfo info;
+        private final BiFunction<Period, Period, Period> operation;
+
+        IntervalArithmeticScalar(String operator, String name) {
+            this.info = new FunctionInfo(
+                new FunctionIdent(name, Arrays.asList(DataTypes.INTERVAL, DataTypes.INTERVAL)), DataTypes.INTERVAL);
+            switch (operator) {
+                case "+":
+                    operation = Period::plus;
+                    break;
+                case "-":
+                    operation = Period::minus;
+                    break;
+                default:
+                    operation = (a,b) -> {
+                        throw new IllegalArgumentException("Unsupported operator for interval " + operator);
+                    };
+            }
+        }
+
+        @Override
+        public Period evaluate(TransactionContext txnCtx, Input<Object>... args) {
+            Period fst = (Period) args[0].value();
+            Period snd = (Period) args[1].value();
+
+            if (fst == null || snd == null) {
+                return null;
+            }
+            return operation.apply(fst, snd);
+        }
+
+        @Override
+        public FunctionInfo info() {
+            return this.info;
+        }
+    }
+
+    private static final class IntervalTimestampScalar extends Scalar<Long, Object> {
+
+        private final BiFunction<DateTime, Period, DateTime> operation;
+        private final FunctionInfo info;
+        private final boolean firstValueIsInterval;
+
+        IntervalTimestampScalar(String operator, String name, DataType firstType, DataType secondType, DataType returnType) {
+            this.info = new FunctionInfo(new FunctionIdent(name, Arrays.asList(firstType, secondType)), returnType);
+            this.firstValueIsInterval = ArithmeticFunctionResolver.isInterval(firstType);
+
+            switch (operator) {
+                case "+":
+                    operation = DateTime::plus;
+                    break;
+                case "-":
+                    if (ArithmeticFunctionResolver.isInterval(firstType)) {
+                        throw new IllegalArgumentException("Unsupported operator for interval " + operator);
+                    }
+                    operation = DateTime::minus;
+                    break;
+                default:
+                    operation = (a,b) -> {
+                        throw new IllegalArgumentException("Unsupported operator for interval " + operator);
+                    };
+            }
+        }
+
+        @Override
+        public Long evaluate(TransactionContext txnCtx, Input<Object>... args) {
+            final Long timestamp;
+            final Period period;
+
+            if (firstValueIsInterval) {
```

Comment:
I had something like

```
period = (Period) args[periodIdx].value();
timestamp = (Long) args[tsIdx].value();
```

in mind. So that there is no additional branching within `evaluate, other than the `null` checks.

Reply:
👍

Reply:
`null + interval '1 second'` yields to `io.crate.exceptions.ConversionException: Cannot cast PT1S to type double precision` without an UNDEFINED type.

Reply:
Ah okay, let's keep it then, I guess it's because https://github.com/crate/crate/pull/8997/files#diff-d4be2122aeae9ba7a07ccf87d4e864b5R181 won't match then and it tries to cast it to `[double, interval]` for whatever reason.

Reply:
Also `interval '1 second' + null` yields to `Cannot cast PT1S to type [double precision, real, bigint, integer, smallint, char, undefined]` without an INTERVAL type defined. I could cover all these cases in `ArithmeticFunctionResolver.getSignature()`, but what would be the benefit of that ?